OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+/** \file
+* This file defines the  DataTable class, which is used by OpenSim to provide  *
+* an in-memory container for data access and manipulation.                     *
+*/
+
+#ifndef OPENSIM_DATA_TABLE_H_
+#define OPENSIM_DATA_TABLE_H_
+
+// Standard headers.
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+#include <string>
+#include <type_traits>
+#include <limits>
+// Non-standard headers.
+#include <SimTKcommon.h>
+
+
+namespace OpenSim {
+  class AbstractDataTable;
+  template<typename ET> class DataTable_;
+
+  using DataTable = DataTable_<SimTK::Real>;
+
+  /// Enum to specify if the the input iterator is traversing data row-wise or
+  /// col-wise.
+  /// clang3.6 crashes if this is turned to a "enum class". Until it is fixed,
+  /// use a pre c++11 enum.
+  enum InputItDir{
+    ROWWISE, 
+    COLWISE
+  };
+
+  /// Bind two DataTables by row.
+  template<typename ET>
+  DataTable_<ET> rbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  /// Bind two DataTables by col.
+  template<typename ET>
+  DataTable_<ET> cbindDataTables(const DataTable_<ET>& dt1, 
+                                 const DataTable_<ET>& dt2);
+
+  // Exceptions.
+  class ColumnDoesNotExist;
+  class ColumnHasLabel;
+  class ColumnHasNoLabel;
+  class ZeroElements;
+  class InvalidEntry;
+  class MetaDataKeyExists;
+  class MetaDataKeyDoesNotExist;
+  class MetaDataTypeMismatch;
+} // namespace OpenSim
+
+
+/**----------------------------------------------------------------------------*
+* AbstractDataTable is the base-class of all DataTable_ allowing storage of    *
+* DataTable_ of different types to be stored in containers. See Datatable_ for *
+* details on the interface.                                                    *
+*-----------------------------------------------------------------------------*/
+class OpenSim::AbstractDataTable {
+public:
+  AbstractDataTable() = default;
+  AbstractDataTable(const AbstractDataTable&) = delete;
+  AbstractDataTable(AbstractDataTable&&) = delete;
+  virtual std::unique_ptr<AbstractDataTable> clone() const = 0;
+  virtual ~AbstractDataTable() {}
+}; // AbstractDataTable
+
+class OpenSim::ColumnDoesNotExist : public std::runtime_error {
+public:
+  ColumnDoesNotExist(const std::string& expl) : runtime_error(expl) {}
```

Comment:
> If every indent eats 4 spaces. templates and long names will make a lot statements go over 80 chars

One thing that helps is to use shorter names in implementations than in the API. That is one reason to split the declarations from the definitions. Because we depend on our declarations to supply doxygen with nice names, they can be too verbose for convenient coding. But the users should win in the API, while programmers rule in the definitions.